### ISSUE TITLE
F401+F811 false positive

### ISSUE THREAD

**Issue by dalcinl** [charliermarsh/ruff#2248]
#### Minimal reproducer
```python
from foo import Foo
from typing import TypeAlias

class Bar:
    Foo: TypeAlias = Foo
    def bar(self) -> Foo: ...
```

#### Output
```
$ ruff stub.pyi 
stub.pyi:1:17: F401 `foo.Foo` imported but unused
stub.pyi:5:5: F811 Redefinition of unused `Foo` from line 1
Found 2 error(s).
1 potentially fixable with the --fix option.
```

Ruff version: `0.0.227`

**Comment by ngnpope** [comment]
Confirmed with `0.0.236` with a `.pyi` file.
Doesn't occur with a `.py` file unless prepending with `from __future__ import annotations`.

**Comment by charliermarsh** [comment]
Is this, maybe, the same as #1401?

**Comment by dalcinl** [comment]
I believe it is not exactly the same. 
My reproducer would be equivalent to:
```python
from foo import Foo

class Bar:
    Foo = Foo
    def bar(self) -> Foo: ...
```
and `ruff` is happy with this code.

**Comment by dalcinl** [comment]
This is a small variation to my original reproducer:

```python
from typing import TypeAlias
class Foo: ...

class Bar:
    Foo: TypeAlias = Foo
    def bar(self) -> Foo: ...
```
This time Ruff is happy with it.
1. My original reproducer triggers the warnings. 
2. Similar code with a locally defined symbol does not.
3. Removing the `TypeAlias ` annotation silences the warning.
4. Changing the annotation to  `Foo: Type = Foo` also silences the warning.
So far this looks like a subtle bug related to visibility and/or the `TypeAlias` annotation.

**Comment by charliermarsh** [comment]
Yeah this is really tricky.

**Comment by charliermarsh** [comment]
Obligatory link to Tweet: https://twitter.com/charliermarsh/status/1659233087786459139

**Comment by charliermarsh** [comment]
I believe we need to change the logic in this case to eagerly resolve if the binding is already defined, and otherwise, go through standard deferred resolution.
